### DISCUSSION TITLE
Make connection between return non nullable value and input non nullable value

### DISCUSSION THREAD

**Discussion by NN---** [dotnet/csharplang#3518] [Language Ideas]
https://github.com/dotnet/runtime/issues/37189#issuecomment-636317983

It is not possible to specify that the return value will be null only if input is null.
Example:

```cs
namespace System.Collections.Generic
{
    public static class CollectionExtensions
    {
        [return: MaybeNull]
        public static TValue GetValueOrDefault<TKey, TValue>(this IReadOnlyDictionary<TKey, TValue> dictionary, TKey key, [AllowNull] TValue defaultValue) where TKey : notnull
```

It should be defined with some new attribute

```cs
[return: MaybeNull]
[return: NullWhenNull("defaultValue")]
public static TValue GetValueOrDefault<TKey, TValue>(
   this IReadOnlyDictionary<TKey, TValue> dictionary,
   TKey key,
   [AllowNull] TValue defaultValue)
    where TKey : notnull
```

This should mean that passing non null value as `defaultValue` results in non null result.

```cs
Dictionary<int, Func<int, int>> d;

var f = d.GetValueOrDefault(1, x => x + 1);

// Obviously f is not null here, but it is hard to say that without some annotations help.
f();
```

**Comment by yaakov-h** [comment]
It looks like you're after the attribute `[return: NotNullIfNotNull("defaultValue")]`?

For a working example, click [here](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AxAHYCuANuRMOXAARzHW0CwAUOwAIBMAjFwAY6nXgBYA3IOG8AdABFMEAObEA9rEwBjKDIDCqgCZwAgk3K4omKJI5sRANmHc6+ynE0xMq4gFEAHvDElt5Q7ADe7HRRdADanADsIHQAshC4wHAAchTkALqR0XGJdJmqMNmUAJLYpeU5ABQAREbYEBQwAGoQ5KRwjQCU+WzRwgDM0o4AKl099ADicJ3dvQDyCHJwre0APJMA0nC4SHTTy3AAfPUFI1EwABZWdJUASnAQBivE5goeXkwIuF2ByOJxmvXOdAMWk83ggAKQ1xu+0OdAA1ocEcMblEYsZKKoAO4VPKgs6QzZtchLWb9REjAl3RD0ZG4OhJNQwMiURERLHYhKQ6F/OG4GSTAELam9erokGqUgwOhoOFKs79OgAflVszZ5K2VLBcHE0URAF92ObbJxxjwXOFEdbhKIUvUfjD/oDMMQYMdOABWbZen10IPnCEGWl8ui87FRZUIOjYOgAXkhMklhrWG31MHqvGOfhTEMLAGo6Lx+jZY4m85WzRb2EA=).

**Comment by NN---** [comment]
Great !
Thanks.

**Comment by NN---** [comment]
Since null can be due to value is null, it should be a new attribute here: NullIfNullAttribute:

```cs
    [return: MaybeNull]
    [return: NullIfNull("defaultValue")]
    public static TValue GetValueOrDefault<TKey, TValue>(
        this IReadOnlyDictionary<TKey, TValue> dictionary,
        TKey key,
        [AllowNull] TValue defaultValue)
        where TKey : notnull
    {
        return dictionary.TryGetValue(key, out var value) ? value : defaultValue;   
    }
```

**Comment by yaakov-h** [comment]
What's the point when its already annotated with `[MaybeNull]`? That wouldn't solve your problem.

**Comment by NN---** [comment]
Ok.
So we have these options:

1. Value is not nullable hence cannot be null
It means that defaultValue cannot be null either => **Returning non null**

2. Value is nullable => **May return null**


Current [MaybeNull] makes the first to be treated as the second.
So it needs to distinguish between two cases.

What about new attribute like ?

```cs
[return: MaybeNullIfNullableElseNotNull("defaultValue")]
```

**Comment by 333fred** [comment]
I would expect the combination of MaybeNull and NotNullIfNotNull to work. Did it not?

**Comment by yaakov-h** [comment]
NotNullIfNotNull would work, but it’s not valid in this scenario as a dictionary’s value can be null.
